source/Halibut/DataStream.cs

```diff
@@ -20,14 +29,56 @@ public DataStream(long length, Action<Stream> writer)
         {
             Length = length;
             Id = Guid.NewGuid();
-            this.writer = writer;
+            this._writer = writer;
+        }
+        
+        
+        public DataStream(Guid guid, Func<DataStream> delayedStreamCreation)
+        {
+            Id = Guid.NewGuid();
+            _delayedStreamCreation = delayedStreamCreation;
         }
 
         [JsonProperty("id")]
         public Guid Id { get; set; }
 
         [JsonProperty("length")]
-        public long Length { get; set; }
+        public long Length
+        {
+            get
+            {
+                EnsureLengthAndStreamInitializedFromDelayedCreation();
+                return _length;
+            }
+            set
+            {
+                EnsureLengthAndStreamInitializedFromDelayedCreation();
+                _length = value;
+            }
+        }
+
+        [JsonIgnore]
+        private long _length = 0;
+
+        [JsonIgnore]
+        Func<DataStream> _delayedStreamCreation = null;
+
+        private void EnsureLengthAndStreamInitializedFromDelayedCreation()
+        {
+            if (_delayedStreamCreation != null)
+            {
+                try
+                {
+                    var r = _delayedStreamCreation();
+                    this._writer = r.writer;
+                    this._length = r.Length;
```

Comment:
This looks suspiciously like a `DataStream` mapper. If we want to map between two DataStreams, can we do that outside of the dto class, so that we're no mixing concerns?